k2/csrc/fsa_equivalent.cc

```diff
@@ -220,22 +238,43 @@ bool IsRandEquivalent(const Fsa &a, const float *a_weights, const Fsa &b,
     // find out that we don't need that version, we will remove flag
     // `top_sorted` and add requirements as comments in the header file.
     CHECK(top_sorted);
-    double sum_a =
-        ShortestDistance<Type>(a_compose_path, a_compose_weights.data());
-    double sum_b =
-        ShortestDistance<Type>(b_compose_path, b_compose_weights.data());
-    if (!DoubleApproxEqual(sum_a, sum_b)) return false;
+    double max_loglike_path_a =
+        ShortestDistance<kMaxWeight>(a_compose_path, a_compose_weights.data());
+    double max_loglike_path_b =
+        ShortestDistance<kMaxWeight>(b_compose_path, b_compose_weights.data());
+    if (max_loglike_path_a < loglike_cutoff_a &&
+        max_loglike_path_b < loglike_cutoff_b) {
+      // both `a` and `b` have no path within `beam` for symbol sequence `path`
+      continue;
+    } else if (max_loglike_path_a >= loglike_cutoff_a &&
+               max_loglike_path_b >= loglike_cutoff_b) {
+      double loglike_path_a, loglike_path_b;
+      if (Type == kLogSumWeight) {
+        loglike_path_a =
+            ShortestDistance<Type>(a_compose_path, a_compose_weights.data());
+        loglike_path_b =
+            ShortestDistance<Type>(b_compose_path, b_compose_weights.data());
+      } else {
+        loglike_path_a = max_loglike_path_a;
+        loglike_path_b = max_loglike_path_b;
+      }
+      if (!DoubleApproxEqual(loglike_path_a, loglike_path_b)) return false;
```

Comment:
we might want the option to pass the delta into this function.

Reply:
ok, will add one.